### PR TITLE
Improve code readability using `examine` macro

### LINKED ISSUE #6491: Add a wrapper around std::visit to allow the value being matched to be given first
In code that looks like
```rust
std::visit(match {
    /* ... lots of code ... */
}, specimen)
```
I often have to scroll down to the end to see what `specimen` is and then scroll back, which is annoying.

### PR DESCRIPTION
**Pull request by daira**
Improve code readability by replacing all instances of
```rust
std::visit(match {
    ...
}, specimen)
```
with a new `examine` macro:
```rust
examine(specimen, match {
    ...
})
```
With the previous ordering I often needed to read the code non-linearly to see what the specimen was, before being able to understand the code in the match body. (In all cases the specimen was a short expression, but the match body can be very large.)

fixes #6491

### CODE REVIEW

**Review by str4d — COMMENTED** [review]
Generally LGTM, however I am blocking this on #6377 so that we don't interfere with the mergeability of the latter.

File: src/util/match.h
Comment (daira) [reply src/util/match.h:32]: The design of the C preprocessor basically assumed that commas in an expression would always be surrounded by parens, which is mostly true for C. But in C++ code it is common to have commas between braces or angle brackets, which causes the preprocessor to treat that code as multiple arguments when passed to a macro. In particular, uses of `match` with more than one alternative will always have this problem, due to the commas between the alternatives. The errors look something like:
```c++
./zcash/Address.hpp:109:17: error: too many arguments provided to function-like macro invocation
                [&](const libzcash::SaplingPaymentAddress &zaddr) {
                ^
./util/match.h:33:9: note: macro 'examine' defined here
#define examine(specimen, visitor) (std::visit((visitor), (specimen)))
        ^
```

Variadic macros parse their arguments in a way that allows commas within braces or angle brackets, so this just works.

Comment (sellout) [reply src/util/match.h:32]: What about avoiding the macro with
```c++
template<class Variant, class Visitor>
constexpr decltype(auto) examine(Variant&& var, Visitor&& vis) {
    return std::visit(vis, var);
}
```
which should work the same at call sites.

Or even
```c++
template<class Variant, class... Visitors>
constexpr decltype(auto) match_visit(Variant&& var, Visitors&&... vis) {
    return std::visit(match { vis... }, var);
}
```
which would eliminate the nested `match`, like
```c++
    examine(changeAddr,
        [&](const CKeyID& keyId) {
            tChangeAddr = keyId;
        },
        [&](const CScriptID& scriptId) {
            tChangeAddr = scriptId;
        },
        [&](const libzcash::SaplingPaymentAddress& changeDest) {
            saplingChangeAddr = std::make_pair(ovk, changeDest);
        },
        [&](const libzcash::OrchardRawAddress& changeDest) {
            orchardChangeAddr = std::make_pair(ovk, changeDest);
        });
```

Comment (sellout) [reply src/util/match.h:32]: @daira I missed that you had a comment shortly before mine that mentioned a similar approach. You’re right – we can’t infer `R` (until C++20), but using `decltype(auto)` works and is how `visit` is actually implemented in Clang for C++17.

Having `R` in C++20 will be nice, because it should eliminate a lot of the `-> some_type` annotations on the individual closures.

Comment (daira) [reply src/util/match.h:32]: I couldn't get the second variant with `std::visit(match { vis... }, var)` to work, but the first one does. Thanks! Force-pushed to use that.

Comment (daira) [reply src/util/match.h]: Btw, I think that `examine` shouldn't be `constexpr` because it doesn't meet [the requirements for a `constexpr` function](https://en.cppreference.com/w/cpp/language/constexpr#constexpr_function). I don't know why clang wasn't complaining.

**Review by sellout — APPROVED** [review]
utACK

But … can we hold this until after #6377 (so, hopefully tomorrow)? It’s going to conflict on a bunch of stuff, and I’d like to not have to wade through more merge conflicts than necessary. Also, #6377 adds a pile more `visit` calls that could be converted in this PR.

**Comment by daira** [comment]
Moved to draft until after #6377 merges.

**Comment by daira** [comment]
Just the last 4 commits. When #6377 merges this should automatically adjust.

**Review by nuttycom — DISMISSED** [review]
utACK

**Review by nuttycom — DISMISSED** [review]
utACK https://github.com/zcash/zcash/pull/6494/commits/8c518f7284900e555a51e0f97e4b7c253eb304f0

**Review by nuttycom — APPROVED** [review]
re-utACK https://github.com/zcash/zcash/commit/3adf580dc07ddec2e0507ca94ad7538268ead600

**Pull request merged**
